src/Microsoft.ApplicationInsights.AspNetCore/DiagnosticListeners/Implementation/HostingDiagnosticListener.cs

```diff
@@ -0,0 +1,104 @@
+﻿namespace Microsoft.ApplicationInsights.AspNetCore.DiagnosticListeners
+{
+    using System;
+    using Microsoft.ApplicationInsights.AspNetCore.Extensions;
+    using Microsoft.ApplicationInsights.DataContracts;
+    using Microsoft.ApplicationInsights.Extensibility.Implementation;
+    using Microsoft.AspNetCore.Http;
+    using Microsoft.Extensions.DiagnosticAdapter;
+
+    internal class HostingDiagnosticListener: IApplicationInsightDiagnosticListener
+    {
+        private readonly TelemetryClient client;
+        private readonly ContextData<long> beginRequestTimestamp = new ContextData<long>();
+        private readonly string sdkVersion;
+
+        public HostingDiagnosticListener(TelemetryClient client)
+        {
+            this.client = client;
+            this.sdkVersion = SdkVersionUtils.VersionPrefix + SdkVersionUtils.GetAssemblyVersion();
+        }
+
+        public string ListenerName { get; } = "Microsoft.AspNetCore";
+
+        [DiagnosticName("Microsoft.AspNetCore.Hosting.BeginRequest")]
+        public void OnBeginRequest(HttpContext httpContext, long timestamp)
+        {
+            if (this.client.IsEnabled())
+            {
+                httpContext.Features.Set(new RequestTelemetry());
+
+                this.beginRequestTimestamp.Value = timestamp;
+                this.client.Context.Operation.Id = httpContext.TraceIdentifier;
+            }
+        }
+
+        [DiagnosticName("Microsoft.AspNetCore.Hosting.EndRequest")]
+        public void OnEndRequest(HttpContext httpContext, long timestamp)
+        {
+            if (this.client.IsEnabled())
+            {
+                var telemetry = httpContext.Features.Get<RequestTelemetry>();
+
+                telemetry.Duration = new TimeSpan(timestamp - this.beginRequestTimestamp.Value);
```

Comment:
We've had some issue with the accuracy of this, please add a TODO comment to use the StopWatch from/like the base SDK.